Various/buyone_Convert custom actions to Lua ReaScript/BuyOne_Convert custom actions in batch to Lua ReaScripts or dump as Lua code.lua

```diff
@@ -0,0 +1,391 @@
+local targ_file = 'reaper-kb.ini'
+
+	if ReaperKeyMap and ReaperKeyMap ~= '' then -- check if the file exists
+		for f in io.popen('dir \"'..path..sep..'KeyMaps'..sep..ReaperKeyMap..'\" /b'):lines() do
+			if f == ReaperKeyMap then break end
+		end
+		if not f then resp = r.MB('    '..ReaperKeyMap..' file\n\n            was\'t found.\n\n   Switching to '..targ_file,'WARNING',1)
+			if resp == 2 then return end
+		else targ_file = 'KeyMaps'..sep..ReaperKeyMap end
+	end
+
+
+-- Construct a table of individual custom actons strings
+local lines_t = {}
+	for line in io.lines(path..sep..targ_file) do
+		if line:match('^ACT') then -- exclude categories besides actions
+			if flags_t[1] and (line:match('%s(0)%s\"') or line:match('%s(100)%s\"') or line:match('%s(32063)%s\"')) then lines_t[#lines_t+1] = line
+			elseif flags_t[2] and line:match('%s(3206[01])%s\"') then lines_t[#lines_t+1] = line end
+		end
+	end
+
+	-- Sort custom action strings numerically and alphabetically by the first character in the custom action name
+	-- 1. Split each code string in two so that the 2nd one starts with the custom action name, and construct a temp. table where each nested table contains two parts of the custom action string
+	local tmp_t = {}
+		for k,v in next, lines_t do
+		tmp_t[k] = {}
+		tmp_t[k][1] = v:match('(.+Custom: )'); tmp_t[k][2] = v:match('Custom: (.+)')
+		end
+	-- 2. Sort the nested tables by the 2nd key value
+	table.sort(tmp_t, function(a,b) return a[2] < b[2] end)
+	-- 3. Re-initialize the original table in a sorted order, joining the split stings back together
+	local lines_t = {}
+	for k,v in next, tmp_t do
+	lines_t[k] = table.concat(v)
+	end
+
+
+-- Re-order the table by sections: Main, Media Explorer, MIDI Editor, MIDI Events list
+
+-- Construct a temp. table in the pre-define order of the Action list section codes
+local captures_t = {'0','100','32063','32060','32061'}
+local tmp_t = {}
+	for _, sect in ipairs(captures_t) do
+		-- Bring custom actions from the same section together, moving their table values in a predefined order of section codes to a temporary table
+		for _,str in next, lines_t do
+		if str:match('%s('..sect..')%s\"') then tmp_t[#tmp_t+1] = str end
+		end
+	end
+
+
+local lines_t = tmp_t
+local tmp_t = nil
+
+	--if converting to individual LUA files
+local resp = flags_t[3] and r.MB('There\'re '..tostring(#lines_t)..' custom actions to export.','INFO',1)
+		if resp == 2 then return end
+
+
+-- Concatenate the path
+	if subdir ~= '' and subdir ~= 'existing or new; or leave as it is or empty' then
+	-- check if such directory already exists and if not, create
+		for dir in io.popen('dir \"'..path..sep..'Scripts\" /ad'):lines() do
+			if dir == subdir then exists = true break end
+		end
+		if not exists then os.execute('mkdir \"'..path..sep..'Scripts'..sep..subdir..'\"')
+		-- check if the directory has been created
+			for dir in io.popen('dir \"'..path..sep..'Scripts\" /b'):lines() do
```

Comment:
This bit is also vulnerable to command injection and won't work on stock macOS (no `dir` is provided). You can use `RecursiveCreateDirectory` from REAPER for creating directories.

Otherwising, this is correctly packaged. Will merge this after the calls to `os.execute` or `io.popen` are fixed.

---

`os.execute` and `io.popen` send a script to the shell for running commands. The shell is a full scripting language interpreter, so directly injecting outside data (user input, file paths) into a shell script (the "command line") is opening a door for bugs to happen.

Answering your question in a previous PR about "sanitizing": this mean ensuring all characters with special meaning to the current system's shell will be interpreted literally. It's somewhat risky and shell syntax varies from system to system.

Escaping for `sh` and derivatives (macOS and most Linux's default) can be as simple as using single quotes and escaping single quotes within the argument though: `'Lua'\''s : operator is nice'`.

But it's better to avoid invoking the system shell when possible. `reaper.ExecProcess` does the argument splitting itself without going to the shell. It supports ' " and ` for quoting, but has no escape character (meaning some values combining space and 2+ types of quotes are not supported).

Reply:
OK. Could you give an example of the kind of harm which can be caused to the user system via these commands by the user themself? Or do you mean it may happen by accident?

Reply:
> Or do you mean it may happen by accident?

Realistically, the worse that is likely to happen is the feature wouldn't work with some input/filenames.

(Anything can happen though: https://i.imgur.com/hkgL0mB.gif.)

Reply:
OK. Could you give an example of the kind of harm which can be caused to the user system via these commands by the user themself? Or do you mean it may happen by accident due to variations of shell syntax on different systems?